aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +27,348 @@ namespace at {
-  bool TH_TENSOR_APPLY_hasFinished = false;
-  int64_t TH_TENSOR_dim_index = 0;
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar1, tensor1, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar2, tensor2, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar3, tensor3, dim, 1)
-
-  int elements_equal = 1;
-  auto t1_numel = tensor1.numel();
-  auto t2_numel = tensor2.numel();
-  auto t3_numel = tensor3.numel();
-  if(t1_numel!= t2_numel) {
-    elements_equal = 0;
-  } else if(t1_numel != t3_numel) {
-    elements_equal = 0;
+AT_API inline std::string _all_equal_numel_error(at::ArrayRef<Tensor> tensors) {
+  std::ostringstream oss;
+  oss << "inconsistent tensor size, expected ";
+  for (size_t i = 0; i < tensors.size() - 1; i++) {
+    oss << tensors[i].sizes() << ", ";
   }
-  if (elements_equal == 0) {
-    std::ostringstream oss;
-    oss << "inconsistent tensor size, expected " << tensor1.sizes() << ", " << tensor2.sizes() << ", and " << tensor3.sizes()
-        << " to have the same number of elements, but got " << t1_numel << ", " << t2_numel << ", and " << t3_numel << " elements respectively";
-    throw std::runtime_error(oss.str());
+  oss << "and " << tensors[tensors.size() - 1]
+      << " to have the same number of elements, but got ";
+  for (size_t i = 0; i < tensors.size() - 1; i++) {
+    oss << tensors[i].numel() << ", ";
   }
+  oss << "and " << tensors[tensors.size() - 1].numel()
+      << " elements respectively";
+  return oss.str();
+}
 
-  while(!TH_TENSOR_APPLY_hasFinished)
-  {
-    /* Loop through the inner most region of the Tensor */
-    for(; tensor1_i <  tensor1_size && tensor2_i < tensor2_size && tensor3_i < tensor3_size; tensor1_i++, tensor2_i++, tensor3_i++, tensor1_data += tensor1_stride, tensor2_data += tensor2_stride, tensor3_data += tensor3_stride)
-    {
-      op(*tensor1_data, *tensor2_data, *tensor3_data);
+AT_API inline bool _apply_preamble(ArrayRef<Tensor> tensors) {
+  checkBackend("CPU_tensor_apply", tensors, Backend::CPU);
+  if (!_all_equal_numel(tensors))
+    throw std::runtime_error(_all_equal_numel_error(tensors));
+  // An empty tensor has no elements
+  for (auto& t : tensors)
+    if (t.sizes().equals({0}))
+      return false;
+  internal::init_tbb_num_threads();
+  return true;
+}
+
+AT_API inline void iterate_all(){};
+
+template <typename Arg, typename... Args>
+AT_API inline void iterate_all(Arg& iter, Args&... iter_tail) {
+  iterate(iter);
+  iterate_all(iter_tail...);
+}
+
+AT_API inline void forward_all(int64_t offset){};
+
+template <typename Arg, typename... Args>
+AT_API inline void forward_all(int64_t offset, Arg& iter, Args&... iter_tail) {
+  forward(offset, iter);
+  forward_all(offset, iter_tail...);
+}
+
+AT_API inline bool continue_loop() {
+  return true;
+}
+
+template <typename Arg, typename... Args>
+AT_API inline bool continue_loop(Arg& iter, Args&... iter_tail) {
+  return iter.counter_[iter.dim_ - 1] < iter.sizes_[iter.dim_ - 1] &&
+      continue_loop(iter_tail...);
```

Comment:
Do we really need to check this condition on all tensors? We've already checked the have an equal number of elements, so it should only be the case that all of those conditions are false or all are true. Is that correct?

Reply:
Yes, for example one iterator might point to a contiguous tensor and the other to a non-contiguous tensor. The number of elements could still be the same. This happens for Unary Ops when returning a tensor opposed to applying the operation to the tensor in-place. We could definitely add more constraints to the applyX functions to make these checks unnecessary. I've kept them around from the apply macro. I suggest we consider this in a separate PR together with all the preprocessing we might consider to speed up these operations overall.

Reply:
Yeah they should be useless in the fully contiguous case, and given how popular and important it is, it would make sense to template on it.